### PR TITLE
Add update strategy to Metal3DataTemplate

### PR DESCRIPTION
**Pull request by kashifest**
Due to the immutability of the Metal3DataTemplate, it was not possible
to update properly the metadata and network data for a
Metal3MachineTemplate. This commits adds an update strategy to allow
the user to create a new template and transition to it via rolling
upgrade of the CAPI machine deployments or KCP.

### CODE REVIEW

**Comment by maelk** [comment]
Is this still WIP ?

File: design/metadata-handling.md [hunk 1]
```diff
@@ -597,15 +608,21 @@ metadata:
...
     namespace: default
-  Metal3DataClaim:
+ claim:
     name: machine-1
     namespace: default
+  template:
+    name: nodepool-1
+    apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    namespace: default
+    kind: Metal3DataTemplate
```
Comment (maelk) [inline design/metadata-handling.md:625]: to be clear, those changes are not related to the proposal, they only fix the deviation between the implementation and the proposal.

File: design/metadata-handling.md [hunk 2]
```diff
@@ -112,6 +117,11 @@ on Metal3Machine
 - Documentation of this feature would be very important.
 - The controllers must be able to recreate all status objects, i.e. no
   necessary information should be stored in the status without being recoverable
+- Since the BareMetalHost references the secrets and they are used at provisioning
+  time, the secrets cannot be updated to be able to reprovision the node in the
+  exact same state. This means that the data template parts containing the
+  metadata and network data must be immutable. Updates have to be done by creating
+  a new template and referencing it in the new / updated Metal3MachineTemplate.
```
Comment (maelk) [inline design/metadata-handling.md:124]: maybe it could be listed as an alternative that the M3DT could be mutable and reconciled so that the secrets would be updated, and reason why this would be problematic. For example, that the reprovisioning of the node would then lead to a different configuration, and that might be unexpected. And if we allow the mutability without updating the secrets,there is no way to know which version of the template they were created with and what they contain.

File: design/metadata-handling.md [hunk 3]
```diff
@@ -630,16 +654,22 @@ object will be reconciled. There will be two cases:
...
-existing Metal3Data object linked to this Metal3DataTemplate and to get the
-unavailable indexes. The indexes always start from 0 and increment by 1. The
-lowest available index is to be used next. The `dataNames` field contains the
-map of Metal3Machine to Metal3Data and the `indexes` contains the map of
-allocated indexes and claims.
+existing Metal3Data object linked to this Metal3DataTemplate and get the
+unavailable indexes. The Metal3Data objects are linked to a Metal3DataTemplate
+by three ways. Either they directly reference the template in the `template`
+field of the `spec`, or they have the same `templateReference` key as the
+template, or the template's `templateReference` matches the `name` of the
```
Comment (maelk) [inline design/metadata-handling.md:667]: If I understood correctly, the third way is to bridge the gap between the old and new implementations, right ? Maybe that should be specified here, that it's a workaround for the migration ?

File: design/metadata-handling.md [hunk 4]
```diff
@@ -597,15 +608,21 @@ metadata:
     name: nodepool-1
 spec:
   index: 0
+  templateReference: SharedRef
```
Comment (maelk) [inline design/metadata-handling.md:611]: Is there a need to add that to the metal3dataclaim too ? My guess would be no, considering that the claims are not shared between M3DTs as the Metal3Data are, right ?

**Review by maelk — COMMENTED** [review]
This looks good to me overall. Maybe the naming "templateReference" is not optimal since it can be quite obscure when in the m3dt itself, what about just "reference" ? or any other idea ?

File: design/metadata-handling.md (hunk 1 above)
Comment (kashifest) [reply design/metadata-handling.md:625]: Yes. Spot on

File: design/metadata-handling.md (hunk 2 above)
Comment (kashifest) [reply design/metadata-handling.md:124]: Just to be clear, do you mean to say I should point out why M3DT mutability is not a viable option?

File: design/metadata-handling.md (hunk 3 above)
Comment (kashifest) [reply design/metadata-handling.md:667]: True, I can repharse it.

File: design/metadata-handling.md (hunk 4 above)
Comment (kashifest) [reply design/metadata-handling.md:611]: Yes, no need to add it to dataclaim as you correctly pointed out.

**Comment by kashifest** [comment]
> This looks good to me overall. Maybe the naming "templateReference" is not optimal since it can be quite obscure when in the m3dt itself, what about just "reference" ? or any other idea ?

I would rather keep it as templateReference just to be crystal clear since there can be other types of references in future, ofcourse if no strong opinions against it.

**Review by fmuyassarov — COMMENTED** [review]
LGTM

File: design/metadata-handling.md (hunk 2 above)
Comment (maelk) [reply design/metadata-handling.md:124]: yes, in the alternatives section

File: design/metadata-handling.md [hunk 5]
```diff
@@ -96,6 +96,11 @@ and be re-used during upgrade.
 As a user, I want to give my own network configuration for cloud-init as a field
 on Metal3Machine
 
+#### Story 5
+
+As a user using a template for metadata and networking, I want to be able to update
```
[the remark this answers is not in the document]
Comment (kashifest) [reply design/metadata-handling.md]: Thanks. Addressed

File: design/metadata-handling.md (hunk 3 above)
Comment (kashifest) [reply design/metadata-handling.md:667]: Addressed

File: design/metadata-handling.md (hunk 2 above)
Comment (kashifest) [reply design/metadata-handling.md:124]: Addressed

**Comment by kashifest** [comment]
@maelk @fmuyassarov PTAL again. I have addressed the comments

**Review by fmuyassarov — COMMENTED** [review]
/lgtm

**Comment by maelk** [comment]
/approve
Since there does not seem to be interest on this from others than our team (it was opened more than two weeks ago), let's proceed to merge it, it's a relative small change.
